codesyncapp/codesync-vs
# Rename Improver

- Optimized logic to traverse a repo using "[glob](https://www.npmjs.com/package/glob)" instead of "[walk](https://www.npmjs.com/package/walk)" 
    - "walk" was causing _High CPU Usage_ and _Maximun Recursion Depth Error_ in VSCode plugin to traverse over large repos
- Optimized logic to checkForRename using [string-similarity](https://www.npmjs.com/package/string-similarity)
    - Earlier logic was taking ~15s for comparing text of files of ~50KB. With "string-similarity" it is in miliseconds
- Fixed a bug in getting `PotentialMatchingFiles` for rename
    - Earlier it was checking against every file of the repo. Now it considers only those files from shadow repo that possess following properties:
        - Shadow file should not be ignorable from .syncignore
        - Actual file is not present for the shadow file
        - Relative path of shadow file should be present in config file
        - Shadow file should not be a binary file since we are going to match the text of files
        - Shadow file should not be empty
    - Also getting `PotentialMatchingFiles` only once per repo instead of once per file
- Improved logic of ignoring a path from .syncignore
    - Using common function everywhere for "[ignores](https://www.npmjs.com/package/ignore)" which ignores path depending on content of .syncignore
    - Improved logic to ignore DEFAULT-SKIP-DIRECTORIES wherever they are present in the project e.g. `node_modules` etc
    
    
 Testing Instructions
 - Try to connect a large repo and observe CPU usage and VSCode logs for performance issues
 - Try to rename some large file (both for IDE-events and non-IDE events) and make sure it is correctly handled as rename
 - Try to copy large number of files in a Connected Repo and make sure those files are uploaded and there is no error in the Logs for performance. 
 - Use different patterns in `.syncignore` such as `!tests/tests_data`, `coverage/*`, ".git" etc and make sure these patterens are ignored/included correctly by the plugin
